### PR TITLE
eth, les, light: filter on logs only, derive receipts on demand

### PR DESCRIPTION
**Pull request by karalabe**
Alternative solution for https://github.com/ethereum/go-ethereum/pull/16162.

This PR changes the event filtering system to only operate on logs, not on entire receipts. This permits the light client to filter for matches without needing to derive generated receipt fields (which require the block). If a match is found, the full block needs to be pulled in to derive the necessary fields.

**Note, this change is much more involved than my previous PR. Review carefully! And test the **** out of it!**

### CODE REVIEW

File: light/odr_util.go [hunk 1]
```diff
@@ -126,15 +126,48 @@ func GetBlock(ctx context.Context, odr OdrBackend, hash common.Hash, number uint
...
+		r := &ReceiptsRequest{Hash: hash, Number: number}
+		if err := odr.Retrieve(ctx, r); err != nil {
+			return nil, err
+		}
+		receipts = r.Receipts
+	}
+	// Return the logs without deriving any computed fields on the receipts
+	logs := make([][]*types.Log, len(receipts))
+	for i, receipt := range receipts {
+		logs[i] = receipt.Logs
```
Comment (zsfelfoldi) [inline light/odr_util.go]: Here you should also fill in the missing fields of the logs

Comment (karalabe) [reply light/odr_util.go]: > // Return the logs without deriving any computed fields on the receipts

**Comment by karalabe** [comment]
No, the point of this change is to fill in **all** the field when the receipt ends up with the user, and that's what it does. The log retrieval is only used by the filtering, which doesn't need these fields.

**Comment by zsfelfoldi** [comment]
Right, my mistake

**Pull request merged**